src/box/box.cc

```diff
@@ -1913,52 +1926,62 @@ box_wait_limbo_acked(double timeout)
 }
 
 /** Write and process a PROMOTE request. */
-static void
-box_issue_promote(uint32_t prev_leader_id, int64_t promote_lsn)
+static int
+box_issue_promote(int64_t promote_lsn)
 {
+	int rc = 0;
 	struct raft *raft = box_raft();
 	uint64_t term = raft->term;
+	uint64_t promote_term = txn_limbo.promote_greatest_term;
 	assert(raft->volatile_term == term);
 	assert(promote_lsn >= 0);
 
 	txn_limbo_begin(&txn_limbo);
-	/*
-	 * XXX: 'begin' takes time. Conditions could change and it is not
-	 * handled anyhow except for these assertions.
-	 */
-	assert(raft->volatile_term == term);
-	assert(txn_limbo.owner_id == prev_leader_id);
-	(void)prev_leader_id;
-	txn_limbo_write_promote(&txn_limbo, promote_lsn, term);
-	txn_limbo_commit(&txn_limbo);
 
+	rc = box_check_election_term_intact(term);
+	if (rc != 0)
+		goto end;
+
+	rc = box_check_promote_term_intact(promote_term);
+	if (rc != 0)
+		goto end;
```

Comment:
This doesn't matter now, but semantically you should call `txn_limbo_rollback`, not `commit`, on an error.
Now they do the same thing, unlock the latch. But someday they might become different.

Reply:
Fixed

Reply:
You forgot to do the same for box_issue_demote

Reply:
My bad, thanks for noticing! Fixed.